pkg/agent/agentserver/server.go

```diff
@@ -196,6 +215,68 @@ func (s *ProxyServer) serveSend(stream agent.ProxyService_ProxyServer, sendCh <-
 	}
 }
 
+func (s *ProxyServer) validateAuthToken(token string) error {
+	trReq := &authv1.TokenReview{
+		Spec: authv1.TokenReviewSpec{
+			Token:     token,
+			Audiences: []string{s.AgentAuthenticationOptions.AuthenticationAudience},
+		},
+	}
+	r, err := s.AgentAuthenticationOptions.KubernetesClient.AuthenticationV1().TokenReviews().Create(trReq)
+	if err != nil {
+		return fmt.Errorf("Failed to authenticate request. err:%v", err)
+	}
+
+	if r.Status.Error != "" {
+		return fmt.Errorf("lookup failed: %s", r.Status.Error)
+	}
+
+	if !r.Status.Authenticated {
+		return fmt.Errorf("lookup failed: service account jwt not valid")
+	}
+
+	// The username is of format: system:serviceaccount:(NAMESPACE):(SERVICEACCOUNT)
+	parts := strings.Split(r.Status.User.Username, ":")
+	if len(parts) != 4 {
+		return fmt.Errorf("lookup failed: unexpected username format")
+	}
+	// Validate the user that comes back from token review is a service account
+	if parts[0] != "system" || parts[1] != "serviceaccount" {
+		return fmt.Errorf("lookup failed: username returned is not a service account")
+	}
+
+	ns := parts[2]
+	sa := parts[3]
+	if s.AgentAuthenticationOptions.AgentNamespace != ns {
+		return fmt.Errorf("lookup failed: incoming request from %q namespace. Expected %q", ns, s.AgentAuthenticationOptions.AgentNamespace)
+	}
+
+	if s.AgentAuthenticationOptions.AgentServiceAccount != sa {
+		return fmt.Errorf("lookup failed: incoming request from %q service account. Expected %q", sa, s.AgentAuthenticationOptions.AgentServiceAccount)
+	}
+
+	return nil
+}
+
+func (s *ProxyServer) authenticateAgent(ctx context.Context) error {
```

Comment:
Can you rename this to authenticateAgentViaToken or something similar?

Reply:
done